src/core/qgstracer.cpp

```diff
@@ -0,0 +1,676 @@
+/***************************************************************************
+  qgstracer.cpp
+  --------------------------------------
+  Date                 : January 2016
+  Copyright            : (C) 2016 by Martin Dobias
+  Email                : wonder dot sk at gmail dot com
+ ***************************************************************************
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+
+#include "qgstracer.h"
+
+#include "qgsgeometry.h"
+#include "qgsgeometryutils.h"
+#include "qgslogger.h"
+#include "qgsvectorlayer.h"
+
+#include <queue>
+#include <vector>
+
+typedef std::pair<int, double> queue_item; // first = vertex index, second = distance
+
+// utility comparator for queue items based on distance
+struct comp
+{
+  bool operator()( const queue_item &a, const queue_item &b )
+  {
+    return a.second > b.second;
+  }
+};
+
+
+// TODO: move to geometry utils
+double distance_2d( const QgsPolyline& coords )
+{
+  int np = coords.count();
+  if ( np == 0 )
+    return 0;
+
+  double x0 = coords[0].x(), y0 = coords[0].y();
+  double x1, y1;
+  double dist = 0;
+  for ( int i = 1; i < np; ++i )
+  {
+    x1 = coords[i].x();
+    y1 = coords[i].y();
+    dist += sqrt(( x1 - x0 ) * ( x1 - x0 ) + ( y1 - y0 ) * ( y1 - y0 ) );
+    x0 = x1;
+    y0 = y1;
+  }
+  return dist;
+}
+
+
+// TODO: move to geometry utils
+double closest_segment( const QgsPolyline& pl, const QgsPoint& pt, int& vertexAfter, double epsilon )
+{
+  double sqrDist = std::numeric_limits<double>::max();
+  const QgsPoint* pldata = pl.constData();
+  int plcount = pl.count();
+  double prevX = pldata[0].x(), prevY = pldata[0].y();
+  double segmentPtX, segmentPtY;
+  for ( int i = 1; i < plcount; ++i )
+  {
+    double currentX = pldata[i].x();
+    double currentY = pldata[i].y();
+    double testDist = QgsGeometryUtils::sqrDistToLine( pt.x(), pt.y(), prevX, prevY, currentX, currentY, segmentPtX, segmentPtY, epsilon );
+    if ( testDist < sqrDist )
+    {
+      sqrDist = testDist;
+      vertexAfter = i;
+    }
+    prevX = currentX;
+    prevY = currentY;
+  }
+  return sqrDist;
+}
+
+/////
+
+/** Simple graph structure for shortest path search */
+struct QgsTracerGraph
+{
+  struct E  // bidirectional edge
+  {
+    //! vertices that the edge connects
+    int v1, v2;
+    //! coordinates of the edge (including endpoints)
+    QVector<QgsPoint> coords;
```

Comment:
I'm wondering... instead of storing a QVector<QgsPoint> as the edge, would it be better to store a QgsCurveV2? This would allow storage of curved edges.